google-cloud-firestore/lib/google/cloud/firestore/query.rb

```diff
@@ -636,13 +668,26 @@ def default_filter
         end
 
         def order_direction direction
-          if direction.to_s.downcase.start_with? "a"
-            :ASCENDING
-          elsif direction.to_s.downcase.start_with? "d"
-            :DESCENDING
-          else
-            :DIRECTION_UNSPECIFIED
+          return :DESCENDING if direction.to_s.downcase.start_with? "d".freeze
+          :ASCENDING
+        end
+
+        def conditionally_add_order_by_name new_query
+          last_direction = :ASCENDING
+
+          new_query.order_by.each do |order_by|
+            return false if order_by.field.field_path == "__name__".freeze
+            last_direction = order_by.direction
           end
+
+          new_query.order_by << StructuredQuery::Order.new(
+            field: StructuredQuery::FieldReference.new(
+              field_path: "__name__".freeze
+            ),
+            direction: last_direction
+          )
+
```

Comment:
Please take a look how this is done in the Node or Java client. You also need to add an order by constraint if there are any inequalities. 

If you change:
`query.where('foo', '<=', 'bar')`
to
`query.where('foo', '<=', 'bar').orderBy('__name__')`
the query becomes invalid. Instead it needs to be:
`query.where('foo', '<=', 'bar').orderBy('foo').orderBy('__name__')`

Reply:
This PR has been updated and should implement the logic as you have explained it.